File: flask_app/app/controllers/discount_usage_controller.py
from flask import request
from flask_jwt_extended import jwt_required, get_jwt_identity
from app.services.discount_usage_service import DiscountUsageService
from app.utils.response import success_response, error_response
from app.utils.decorators import admin_required


class DiscountUsageController:

    @staticmethod
    @admin_required()
    def get_all():
        page = request.args.get("page", default=1, type=int)
        per_page = request.args.get("per_page", default=10, type=int)

        filters = {
            "discount_id": request.args.get("discount_id"),
            "user_id": request.args.get("user_id"),
        }

        filters = {k: v for k, v in filters.items() if v is not None}

        usages, total = DiscountUsageService.get_all(page, per_page, filters=filters)
        return success_response(
            "Discount usages retrieved",
            {
                "items": [u.to_dict() for u in usages],
                "total": total,
                "page": page,
                "per_page": per_page,
            },
        )

    @staticmethod
    @jwt_required()
    def get_one(usage_id):
        usage = DiscountUsageService.get_by_id(usage_id)
        if not usage:
            return error_response("Discount usage not found", 404)
        return success_response("Discount usage found", usage.to_dict())

    @staticmethod
    @jwt_required()
    def get_my_usages():
        current_user_id = get_jwt_identity()
        usages = DiscountUsageService.get_by_user(current_user_id)
        return success_response(
            "Your discount usages retrieved", [u.to_dict() for u in usages]
        )

File: springboot_api/src/main/java/com/example/membership_api/controller/DiscountUsageController.java
package com.example.membership_api.controller;

import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.example.membership_api.annotations.AdminOnly;
import com.example.membership_api.model.DiscountUsage;
import com.example.membership_api.service.DiscountUsageService;
import com.example.membership_api.utils.JwtUtil;
import com.example.membership_api.utils.ResponseUtil;

import jakarta.servlet.http.HttpServletRequest;

@RestController
@RequestMapping("/discount-usages")
public class DiscountUsageController {

    @Autowired
    private DiscountUsageService service;

    @GetMapping
    @AdminOnly
    public ResponseEntity<?> getAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int perPage,
            @RequestParam(required = false) String userId,
            @RequestParam(required = false) String discountId
    ) {
        Map<String, Object> data = service.getAll(page, perPage, userId, discountId);
        return ResponseUtil.success("Discount usages retrieved", data);
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getOne(@PathVariable String id) {
        DiscountUsage usage = service.getById(id);
        if (usage == null) {
            return ResponseUtil.error("Discount usage not found", 404);
        }
        return ResponseUtil.success("Discount usage found", usage);
    }

    @PostMapping
    public ResponseEntity<?> create(@RequestBody DiscountUsage discountUsage) {
        DiscountUsage created = service.create(
                discountUsage.getDiscountId(),
                discountUsage.getUserId(),
                discountUsage.getAppliedOn(),
                discountUsage.getEntity(),
                discountUsage.getDiscountAmount()
        );
        return ResponseUtil.success("Discount usage created", created);
    }

    @GetMapping("/me")
    public ResponseEntity<?> getMyUsage(HttpServletRequest request, @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int perPage
            ) {
        String userId = JwtUtil.extractUserIdFromRequest(request);
        List<DiscountUsage> usage = service.getByUser(page, perPage, userId);
        if (usage == null) {
            return ResponseUtil.error("Discount usage not found", 404);
        }
        return ResponseUtil.success("Discount usage found", usage);
    }
}
